scripts/parse-vcf/parse_vcf.py

```diff
@@ -0,0 +1,408 @@
+    query_result = subprocess.run(
+        [
+            "bcftools",
+            "query",
+            path,
+            "-H",
+            "-r",
+            "{0}:{1}-{2}".format(*config),
+            "--exclude",
+            "INFO/AF>0.02",
+            "-f",
+            "%CHROM\t%POS\t%REF\t%ALT\t%INFO/vep\t%INFO/nhomalt\t%INFO/AC\t%INFO/AF\n",
+        ],
+        capture_output=True,
+        timeout=90,  # prevent process from hanging in case of connectivity issue
+    )
+
+    if query_result.stderr:
+        raise Exception(f"bcftools error!: {query_result.stderr}")
+
+    return query_result.stdout
+
+
+def parse_vcf_response(vcf: bytes):
+    callset_df = cast(
+        pd.DataFrame,
+        pd.read_csv(
+            BytesIO(vcf),
+            sep="\t",
+            na_values=".",
+            dtype={
+                "# [1]CHROM": str,
+                "[2]POS": int,
+                "[3]REF": str,
+                "[4]ALT": str,
+                "[5]vep": str,
+                "[6]nhomalt": int,
+                "[7]AC": int,
+                "[8]AF": float,
+            },
+        ),
+    )
+    """ remove bcftools' brackets from headers and convert to lower case """
+    callset_df.rename(
+        columns=lambda header: re.sub(r"(\[.+\])|\# | \W", "", header).lower(),
+        inplace=True,
+    )
+    callset_df = parse_vep(callset_df)
+
+    callset_df = callset_df.drop_duplicates()
+
+    callset_df["af"] = callset_df["af"].fillna(0)
+
+    return callset_df
+
+
+def parse_vep(df: pd.DataFrame):
+    # FYI
+    VEP_HEADERS = "Allele,Consequence,IMPACT,SYMBOL,Gene,Feature_type,Feature,BIOTYPE,EXON,INTRON,HGVSc,HGVSp,cDNA_position,CDS_position,Protein_position,Amino_acids,Codons,Existing_variation,ALLELE_NUM,DISTANCE,STRAND,FLAGS,VARIANT_CLASS,MINIMISED,SYMBOL_SOURCE,HGNC_ID,CANONICAL,TSL,APPRIS,CCDS,ENSP,SWISSPROT,TREMBL,UNIPARC,GENE_PHENO,SIFT,PolyPhen,DOMAINS,HGVS_OFFSET,GMAF,AFR_MAF,AMR_MAF,EAS_MAF,EUR_MAF,SAS_MAF,AA_MAF,EA_MAF,ExAC_MAF,ExAC_Adj_MAF,ExAC_AFR_MAF,ExAC_AMR_MAF,ExAC_EAS_MAF,ExAC_FIN_MAF,ExAC_NFE_MAF,ExAC_OTH_MAF,ExAC_SAS_MAF,CLIN_SIG,SOMATIC,PHENO,PUBMED,MOTIF_NAME,MOTIF_POS,HIGH_INF_POS,MOTIF_SCORE_CHANGE,LoF,LoF_filter,LoF_flags,LoF_info\n"
+    CANONICAL_INDEX = 26
+    BIOTYPE_INDEX = 7
+    TARGET_INDEXES = {4: "gene", 6: "transcript", 12: "cdna", 15: "amino_acids"}
+
+    def vep_to_array(vep: str):
+        rows = vep.split(",")
+        target_vep_row = None
+        for row in rows:
+            is_canonical = False
+            is_protein_coding = False
+            vals = row.split("|")
+            for i, val in enumerate(vals):
+                if i == CANONICAL_INDEX and val == "YES":
+                    is_canonical = True
+                if i == BIOTYPE_INDEX and val == "protein_coding":
+                    is_protein_coding = True
+            if is_canonical and is_protein_coding:
```

Comment:
Can you elaborate on the significance behind this logic? Is VEP only reliable for canonical and protein coding genes?

Reply:
There will be a VEP entry for every transcript, of which there might be several for a given SNP --- for now we're only interested in canonical/protein coding.